### ISSUE TITLE
[clang-tidy] readability-use-concise-preprocessor-directives, take consistency into account

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#155680]
Consider this example:

```cpp
#if defined(FOO)
int x;
elif defined(BAR)
int y;
else
int z;
```

Here we use `if defined` because we don't have `elifdef` until C++23, so we want to keep consistency. However the check warns about it.

Not sure how difficult it would be to implement or if it's worth at all, but would be nice if the check was smart enough to understand this use case (possibly under an option). Otherwise if we think it's too niche of a use case I can just close this and NOLINT it :)

**Comment by LegalizeAdulthood** [comment]
IMO, this is a false positive

**Comment by carlosgalvezp** [comment]
Good points! For me personally I would say that given an `#if - #endif` block (including nested #if - #endif), best would be if the whole block had the same style, concise or not. So in your last example I would vote for keeping `defined` in both branches.

Then comes the question of whole-file consistency, e.g. 

```cpp
#ifdef FOO
#endif

#if defined BAR
#else if defined BAZ
#endif

```

Some people might argue it's inconsistent that both blocks don't follow the same style. I don't have a very strong opinion, perhaps leaning towards just keeping the whole file consistent. The exception to this rule could be header guards, since it's pretty standard to use the concise form. 

This is a very subjective topic so I'm sure there's no right answer, probably best is to have options for people to tweak. But that increases the check's complexity and testing of course, so we need to weight that effort in.

**Comment by localspook** [comment]
For whole-file consistency, I'm a bit skeptical. I can imagine a large file all using concise style, then someone (pre-C++23) goes and adds one `#elif defined(FOO)`. Suddenly, all the directives in the file need to change to "expanded" style, and the resulting diff has a lot of noise. I imagine many people wouldn't want that tradeoff. (Although if there _are_ people that want it, we can of course implement it.)
